lib/math.ex

```diff
@@ -472,4 +472,65 @@ defmodule Math do
   """
   @spec atanh(x) :: float
   defdelegate atanh(x), to: :math
+
+  @doc """
+  Computes the modular multiplicatibe inverse of `a` under modulo `m`
+
+  In other words, given integers `a` and `m` calculate a value `b` for `ab = 1 (mod m)`
+
+  Returns an `{:ok, b}` tuple or `{:error, :not_coprime}` when `b` cannot be calculated for the inputs.
+
+  ## Examples
+
+      iex> Math.mod_inv 3, 11
+      {:ok, 4}
+      iex> Math.mod_inv 10, 17
+      {:ok, 12}
+      iex> Math.mod_inv 123, 455
+      {:ok, 37}
+      iex> Math.mod_inv 123, 456
+      {:error, :not_coprime}
+
+  """
+  @spec mod_inv(a :: integer, m :: integer) :: {:ok, integer} | {:error, :not_coprime}
+  def mod_inv(a, m)
+
+  def mod_inv(a, m) do
+    1..(m - 1)
+    |> Enum.find(:not_coprime, &(rem(a * &1, m) == 1))
```

Comment:
A good initial algorithm, but we can indeed probably find a better approach. :slightly_smiling_face: 


And as a rule of thumb, if you try to pipe into an anonymous function, then it nearly always is a better idea to create a private named helper function for it.

Reply:
> A good initial algorithm, but we can indeed probably find a better approach.

I'm planning on writing a body recursive variant modeled after [this C++ code](https://www.techiedelight.com/extended-euclidean-algorithm-implementation/) as well as a tail-end recursive variant based on [this (scroll to bottom)](https://brilliant.org/wiki/extended-euclidean-algorithm/). I'm going to time them and try to play around to see if there is significant difference or special use cases (size of inputs) for certain ones.

I know in the Elixir/Erlang world we like our tail-end recursion, but I've seen several SO and blogs about there being cases where body recursive is faster. Not sure if you have any more insight.

> And as a rule of thumb, if you try to pipe into an anonymous function, then it nearly always is a better idea to create a private named helper function for it.

Is this based on performance or readability? Either way, I didn't put too much effort into this version since I knew that I'd be tossing it eventually. Just wanted to get the tests and other things set up. :sweat_smile:

Reply:
> Is this based on performance or readability?

Readability, although it will not hamper performance in any way (the Elixir/Erlang compilers inline private functions).

Reply:
> Either way, I didn't put too much effort into this version since I knew that I'd be tossing it eventually. Just wanted to get the tests and other things set up. sweat_smile

Of course :grinning: !

Reply:
I suggest looking on [RosettaCode](https://rosettacode.org/wiki/Modular_inverse) for inspiration. There are some great algorithms (both iterative, body-recursive and tail-recursive) on there. I suspect that in this case the tail recursive version will be faster or the two versions will be equally fast (with the body-recursive version taking more stack memory during execution) because we are manipulating a small data structure (an integer) rather than e.g. a list that is growing and needs to be reversed at the end. But we should maybe benchmark this? :grin: